### ISSUE TITLE
List all communities on a remote instance

### ISSUE THREAD

**Issue by Falmarri** [LemmyNet/lemmy#3105]
- [ x ] Did you check to see if this issue already exists?
- [ x ] Is this only a single feature request? Do not put multiple feature requests in one issue.
- [ x ] Is this a question or discussion? Don't use this, use https://lemmy.ml/c/lemmy_support.
- [ x ] Is this a UI / front end issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo.

**Describe the feature request below**

On the communities list (for example https://lemmy.world/communities ) you can choose local or remote. But there's no way to see the list of communities on a _specific_ instance. You should be able to list the communities on some specific instance. The only way it seems to be able to do that is to go to that instance, but then the list isn't complete because I believe they're allowed to gate the display on being logged in (correct me if I'm wrong there). But regardless, you shouldn't have to go off instance to see this list.

Maybe something like this:

![lemmy_community_prototype](https://github.com/LemmyNet/lemmy/assets/463948/e4e0c7fa-86d3-472c-b3d3-1ca8e927d929)

Or possibly a different url, like here:

Have an additional link, instead of to https://lemmy.ml/c/asklemmy (IMO this should probably link to https://lemmy.world/c/asklemmy@lemmy.ml anyway), but link to https://lemmy.world/c/@lemmy.ml or something. Or maybe https://lemmy.world/c/*@lemmy.ml should be a valid link that will wildcard search.

**Comment by krestenlaust** [comment]
While this wouldn't require too much effort to implement, it would be nugatory, as one can simply open the local view on another instance. On the other hand, I like the idea of not having to leave one's own instance.

**Comment by jordan1776** [comment]
> While this wouldn't require too much effort to implement, it would be nugatory, as one can simply open the local view on another instance. On the other hand, I like the idea of not having to leave one's own instance.

I would disagree, features like these are necessary to make a cohesive experience across Lemmy, Anytime where a user has to leave the website or make a new tab where they are logged out is a design failure. I mentioned in an issue on lemmy-ui [#1298
](https://github.com/LemmyNet/lemmy-ui/issues/1298) a similar idea with more filters and a sorting ability.

**Comment by krestenlaust** [comment]
Yes, but what is the REAL use of viewing communities by instance. I believe, that in reality an instance shouldn't be a term really used by the end-user, instead you have users and communities, that's it. In the same way that you don't ever stop to consider what implications it has that the email you're sending something to, is hosted at gmail.com instead of yahoo.com.

Instead the user should have an ability to search all communities (that the user's instance federates with) at once. Similar to [https://browse.feddit.de](https://browse.feddit.de).

I don't think it's going to difficult to implement, and it's such a minor feature, that it wouldn't make any big difference, but I'll still make the argument, that the end-user shouldn't concern themselves with instances.

**Comment by jordan1776** [comment]
> Yes, but what is the REAL use of viewing communities by instance. I believe, that in reality an instance shouldn't be a term really used by the end-user, instead you have users and communities, that's it. In the same way that you don't ever stop to consider what implications it has that the email you're sending something to, is hosted at gmail.com instead of yahoo.com.

Well that depends on how instances will be used as more users join Lemmy. I am in the camp that an instance should basically be treated like a cool handle, like the way email is now. But the way I see it now is it may go in the direction of a few major instances, and many smaller instances. The major instances would have their big general communities with broad appeal (tech, news, jokes, sports). The smaller ones would totally forgo these generals for exclusively hosting communities of niche content which appeals to the same demographic. 

And this is were being able to filter by instance becomes handy. Lets go with a made up example, instance 'techyhobbies.com' which has communities around keyboards, headphones, and fountain pens. Someone who subscribes to 2 of the communities might not realize that the 3rd exists (only a minor interest, not discovered by their home instance). But if they were 1 dropdown away from finding it (neatly grouping it with other communities they like) instead of the current ways, it would make user discovery of small communities much easier.

**Comment by RyantHults** [comment]
> Yes, but what is the REAL use of viewing communities by instance. I believe, that in reality an instance shouldn't be a term really used by the end-user, instead you have users and communities, that's it. In the same way that you don't ever stop to consider what implications it has that the email you're sending something to, is hosted at gmail.com instead of yahoo.com.

I have been seeing this exact sentiment quite a bit. Especially for when the smaller instances are getting off the ground. Right now, a new user joining a new instance will be presented with a completely blank site until they (or the site-admin) search for a community with the right link/syntax.

**Comment by krestenlaust** [comment]
> And this is were being able to filter by instance becomes handy. Lets go with a made up example, instance 'techyhobbies.com' which has communities around keyboards, headphones, and fountain pens. Someone who subscribes to 2 of the communities might not realize that the 3rd exists (only a minor interest, not discovered by their home instance). But if they were 1 dropdown away from finding it (neatly grouping it with other communities they like) instead of the current ways, it would make user discovery of small communities much easier.

Maybe, it makes more sense to simply have a search feature, where you can start your search with an `@` and it just brings up all communities associated with that instance?

Another solution might be to have dedicated pages for an instance, where you can read the rules and view all communities. Similar to how you can open a community-page and view all posts.

**Comment by krestenlaust** [comment]
> I have been seeing this exact sentiment quite a bit. Especially for when the smaller instances are getting off the ground. Right now, a new user joining a new instance will be presented with a completely blank site until they (or the site-admin) search for a community with the right link/syntax.

> Even just an option for instance owners to pull down a list of federated communities for users to check out would be massively helpful for user adoption imo.

Aren't these separate issues? Being able to view communities by instance isn't going to resolve this. The best way to start out is to just use the 'all' feed, and subscribe to relevant communities.

**Comment by jordan1776** [comment]
>Maybe, it makes more sense to simply have a search feature, where you can start your search with an @ and it just brings up all communities associated with that instance?

That might be one way of doing it. I believe reddit's search includes typed operators for filtering by subreddit, author, link, etc. Although direct buttons/dropdowns might be easier on users without keyboards, if it should remain as a feature for power users, this would suffice. I still lean towards buttons/dropdowns for newbies and half awake who just want to press buttons until they find something that looks interesting.

>Aren't these separate issues? Being able to view communities by instance isn't going to resolve this. The best way to start out is to just use the 'all' feed, and subscribe to relevant communities.

What about for communities that don't hit the 'all' feed beyond the 'new' sort?

**Comment by krestenlaust** [comment]
I just realized this issue was created on `lemmy` and not `lemmy-ui`. It should probably be moved.

> That might be one way of doing it. I believe reddit's search includes typed operators for filtering by subreddit, author, link, etc. Although direct buttons/dropdowns might be easier on users without keyboards, if it should remain as a feature for power users, this would suffice. I still lean towards buttons/dropdowns for newbies and half awake who just want to press buttons until they find something that looks interesting.

Both it is, then.

> What about for communities that don't hit the 'all' feed beyond the 'new' sort?

This issue won't solve that problem anyway. Getting users onboarded is a whole separate issue.

**Comment by jordan1776** [comment]
>I just realized this issue was created on lemmy and not lemmy-ui. It should probably be moved.

Agreed. Only thing is with communities that have been undiscovered by the instance being searched from, should those show up as well? I think yes, but how could it be done without straining the instance being searched? There looks to be some talk about indexing on #3192 and #2951 .

**Comment by krestenlaust** [comment]
While those are interesting issues for sure, I don't believe they make this issue any more backend related :)

**Comment by Rebouz** [comment]
petition to merge this with #3207 as this sounds like two sides of the same coin for me, maybe just two different views for it

and then if this is frontend related maybe this needs to be closed; if so we can just pick it up where we left of in a new issue on the [frontend](https://github.com/LemmyNet/lemmy-ui). I don't wanna ping dessalines because he's very busy but I'm bad with rust so I'm not sure myself.

also wanna bump this because I had this exact problem. I heard there was a new instance but didn't know which community to join to get on topic, and then it would have probably taken me 10 minutes to figure out how to join that community from my instance.

**Comment by tgxn** [comment]
in the meantime, I have added an instance-filter on https://lemmyverse.net/communities
I also publish all my crawled data here https://data.lemmyverse.net/ which could be used to drive this feature, there are several mobile apps that already use this

speaking to this feature being integrated into Lemmy, I feel like there could be some potential issues to think about depending on how it is implemented.

1. if it were implemented as a simple crawler job on each instance - it could put excessive load on the whole federated network (if the data is crawled from every connected instance)
2. if it was driven by the local database entries for the remote instance - a server might only show a subset of a remote instances communities that it "knows about" (ie. ones that someone on the instance is already subscribed to)
3. if it were implemented as a direct call to the remote instance - it might put undue load on the "big servers if a lot of instances are performing lookup against them.

I personally thing option 3 could be a good happy medium, and there are definitely mitigations we could take to make these more viable.
